### PR TITLE
Fix completion in indented lines dropping prefix when jedi is disabled

### LINKED ISSUE #14470: Completion in indented lines incorrectly strips the prefix when jedi is disabled
Completion in indented lines incorrectly strips the prefix when jedi is disabled.

Originally reported in Notebook (https://github.com/jupyter/notebook/issues/7397) then JupyterLab (https://github.com/jupyterlab/jupyterlab/issues/16522), but narrowed down to IPython 8.8 / 8.11; https://github.com/ipython/ipython/pull/13943 fixed a related issue but did not cover this one.

### PR DESCRIPTION
**Pull request by krassowski**
Fixes #14470

I would appreciate someone testing this PR.

### CODE REVIEW

File: IPython/core/completer.py [hunk 1]
```diff
@@ -2316,9 +2326,41 @@ def _jedi_matches(
...
+                return _convert_matcher_v1_result_to_v2(
+                    matches, type="attribute", fragment=fragment
+                )
+            except NameError:
+                # catches <undefined attributes>.<tab>
+                matches = []
+                return _convert_matcher_v1_result_to_v2(matches, type="attribute")
+        else:
+            matches = self.global_matches(context.token)
+            return _convert_matcher_v1_result_to_v2(matches, type="variable")
```
Comment (krassowski) [inline IPython/core/completer.py:2356]: TODO: check the exact type name used by jedi to align if we can (is it attribute/variable or something different)?

**Comment by krassowski** [comment]
It would be good to add an E2E test for this one, most unit tests are on lower level functions

File: IPython/core/completer.py [hunk 2]
```diff
@@ -2316,9 +2326,41 @@ def _jedi_matches(
             else:
                 return iter([])
 
+    @context_matcher()
+    def python_matcher(self, context: CompletionContext) -> SimpleMatcherResult:
+        """Match attributes or global python names"""
+        text = context.line_with_cursor
```
Comment (krassowski) [inline IPython/core/completer.py:2332]: TODO: double check if this should be `line_with_cursor` or `token`

File: IPython/core/completer.py [hunk 3]
```diff
@@ -2316,9 +2326,41 @@ def _jedi_matches(
...
+                            is None
+                        )
+                    matches = filter(no__name, matches)
+                return _convert_matcher_v1_result_to_v2(
+                    matches, type="attribute", fragment=fragment
+                )
+            except NameError:
+                # catches <undefined attributes>.<tab>
+                matches = []
+                return _convert_matcher_v1_result_to_v2(matches, type="attribute")
```
Comment (krassowski) [inline IPython/core/completer.py:2353]: Jedi uses `property`, LSP uses `property`/`field`. In Python `property` is a special kind of an attribute, so `attribute` is more accurate.

File: IPython/core/completer.py (hunk 1 above)
Comment (krassowski) [reply IPython/core/completer.py:2356]: Jedi uses `statement`, LSP uses `variable`. I think `variable` is more appropriate here; a variable may however represent a more specialised "thing" such as `module` or `function`. While this is outside of this PR scope, I will add a comment to remember to update it.

Comment (krassowski) [reply IPython/core/completer.py:2356]:
```suggestion
            # TODO: maybe distinguish between functions, modules and just "variables"
            return _convert_matcher_v1_result_to_v2(matches, type="variable")
```

**Comment by krassowski** [comment]
Thank you @spreiter, I fixed this now.

**Comment by spreiter** [comment]
Tried with https://github.com/krassowski/ipython/commit/5827cc8b7a69a67076bbf0c903dfcf0617ca96af , looks good. I can currently not think of any other test...

So all in all, good work from my testing point of view.

**Pull request merged**